comictalker/talkerbase.py

```diff
@@ -0,0 +1,218 @@
+            2 - Network
+            3 - Data
+        desc -- description of the error
+        source -- the name of the source producing the error
+    """
+
+    codes = {
+        1: "General",
+        2: "Network",
+        3: "Data",
+        4: "Other",
+    }
+
+    def __init__(self, source: str = "", code: int = 4, desc: str = "", sub_code: int = 0) -> None:
+        super().__init__()
+        if desc == "":
+            desc = "Unknown"
+        self.desc = desc
+        self.code = code
+        self.code_name = self.codes[code]
+        self.sub_code = sub_code
+        self.source = source
+
+    def __str__(self):
+        return f"{self.source} encountered a {self.code_name} error. {self.desc}"
+
+
+class TalkerNetworkError(TalkerError):
+    """Network class exception for information sources
+
+    Attributes:
+        sub_code -- numerical code for finer detail
+            1 -- connected refused
+            2 -- api key
+            3 -- rate limit
+            4 -- timeout
+    """
+
+    net_codes = {
+        0: "General network error.",
+        1: "The connection was refused.",
+        2: "An API key error occurred.",
+        3: "Rate limit exceeded. Please wait a bit or enter a personal key if using the default.",
+        4: "The connection timed out.",
+        5: "Number of retries exceeded.",
+    }
+
+    def __init__(self, source: str = "", sub_code: int = 0, desc: str = "") -> None:
+        if desc == "":
+            desc = self.net_codes[sub_code]
+
+        super().__init__(source, 2, desc, sub_code)
+
+
+class TalkerDataError(TalkerError):
+    """Data class exception for information sources
+
+    Attributes:
+        sub_code -- numerical code for finer detail
+            1 -- unexpected data
+            2 -- malformed data
+            3 -- missing data
+    """
+
+    data_codes = {
+        0: "General data error.",
+        1: "Unexpected data encountered.",
+        2: "Malformed data encountered.",
+        3: "Missing data encountered.",
+    }
+
+    def __init__(self, source: str = "", sub_code: int = 0, desc: str = "") -> None:
+        if desc == "":
+            desc = self.data_codes[sub_code]
+
+        super().__init__(source, 3, desc, sub_code)
+
+
+# Class talkers instance
+class TalkerBase:
+    """This is the class for mysource."""
+
+    def __init__(self) -> None:
+        # Identity name for the information source etc.
+        self.source_details: SourceDetails | None = None  # Can use this to test if custom talker has been configured
+
+        self.log_func: Callable[[str], None] | None = None
+
+    def set_log_func(self, log_func: Callable[[str], None]) -> None:
+        self.log_func = log_func
+
+    # issueidentifier will set_log_func to print any write_log to console otherwise logger.info is not printed
+    def write_log(self, text: str) -> None:
+        if self.log_func is None:
+            logger.info(text)
+        else:
+            self.log_func(text)
```

Comment:
Instead of having the Talker decide what to do with logs. Lets give it a function that takes a string `def fun(text: str):` that it will use for user output and for debug/other logging we will use the `logger = logging.getLogger(__name__)` defined at the top of the file.

Reply:
I'm sure you're sick of my stupid questions but I'm having trouble getting my head around how this would work? `issueidentifer` is the one deciding what to do as it's the only thing that uses `set_log_func` in talker? That's after `cli`, `volumesectionwindow` or `taggerwindow` tells `issueidentifer` what to do via `set_output_function` it seems?

This is probably really simple and I'm trying to over complicate things?
```
log_func(self, msg: str) -> None:
  print(msg)  # ?
```
That doesn't seem right? Doesn't `set_log_func` need to stay?

Reply:
For the record I don't find any of your questions stupid.

> When you have a class that gets subclassed, like we are doing here, the only thing that is actually accounted for is the types and because this is python types are not actually enforced. Even though we create the initial class and put these two functions in, any subclass can do whatever they would like with it the only thing they have to conform to is the number of arguments. So if they chose to they could override `set_log_func` and make it do something else, now it doesn't make sense to and if they did they would probably be trying to cause issues, so maybe I'm being more paranoid than I should be.
That's my reasoning for wanting to change what's here.

Instead we would pass it into the constructor and set it there.
So the constructor would change to:
```python
def __init__(self, log_func) -> None:
    self.write_log = log_func
```

Looking at it again though I think we should just remove it altogether, anywhere that it is used it's only information that is useful as debug information, it is never something that the user would need to see in normal operation of ComicTagger, the only exception is `search_for_series` but it has it's own callback parameter.

I think it would be better to replace all of the calls to `self.write_log` with `logger.debug` and then remove both of these functions

Reply:
In the talker there is logging along with using an exception. Does it make sense to have the exception handle write to the log or no?
```
            except requests.exceptions.Timeout:
                logger.debug("Connection to " + self.source_name_friendly + " timed out.\n")
                raise TalkerNetworkError(self.source_name_friendly, 4)
```

Reply:
As in having the `TalkerException` have a call to `logger.*` no definitely not, that's what's known as side effects. But having `ComicTalker` log during an exception is fine, but take off the trailing new line and use an fstring